### PR TITLE
libreoffice: fix -qt build

### LINKED ISSUE #165174: `libreoffice-qt` Fails to Build
### Describe the bug
When updating/installing `libreoffice-qt`, it fails with the following error:
```
configure: error: KF5 headers not found.  Please specify the root of your KF5 installation by exporting KF5DIR before running "configure".
error: builder for '/nix/store/nq2n6p077x4fhywaqij9lij7q5v80w3b-libreoffice-7.2.5.2.drv' failed with exit code 1;
       last 10 log lines:
       > checking for QtCore/qconfig.h... yes
       > checking for Qt5 libraries... /nix/store/g6lhjbkcz9nbwlsdd83qsa3rg6i90rs7-qtbase-5.15.3/lib
       > checking for xcb... yes
       > checking for xcb-icccm... yes
       > checking for moc-qt5... no
       > checking for moc... /nix/store/0h41dvl7ipzzm2nj3qfd3vkvaiw6iz9z-qtbase-5.15.3-dev/bin/moc
       > checking for gobject-2.0... yes
       > checking for kf5-config... no
       > checking for KF5 headers... no
       > configure: error: KF5 headers not found.  Please specify the root of your KF5 installation by exporting KF5DIR before running "configure".
       For full logs, run 'nix log /nix/store/nq2n6p077x4fhywaqij9lij7q5v80w3b-libreoffice-7.2.5.2.drv'.
error: 1 dependencies of derivation '/nix/store/7k73impmw6li61ryjn91wix5sjsln2im-libreoffice-7.2.5.2.drv' failed to build
```

### Steps To Reproduce
1. Execute `nix-shell -p libreoffice-qt`
2. Observe build failure

### Expected behavior
libreoffice-qt to install without errors.

### Additional context
KDE Plasma 5.24.0 just landed in `master`, so this might be related to it.

### Notify maintainers
@7c6f434c

### Metadata
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.29, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.0pre20220311_d532269`
 - channels(root): `"nixos"`
 - channels(bill): `"home-manager"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Description of changes

We were doing some rather ugly patching for LO to find the Qt & KDE dependencies which was rather brittle, so instead of trying to fight the build system, use symlinkJoin to set up the dependencies and patch that in.

There is a *lot* of noise due to nixpkgs-fmt. If the LO maintainers insist, I can remove those changes.

A few minor other changes also snuck in: 
 - use buildTargets instead of a manually overridden buildPhase
 - use getDev and getLib instead of referencing outputs directly
 - pull in what we need from lib

Fixes: #165174

Not built yet - doing it on a laptop isn't my idea of fun...

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by peterhoeg** [comment]
@GrahamcOfBorg build libreoffice-fresh libreoffice-qt

**Comment by peterhoeg** [comment]
@GrahamcOfBorg build libreoffice-qt

**Comment by peterhoeg** [comment]
I've built this locally and it works (ofborg times out). So this PR does work - the issue is if @7c6f434c will accept it.

**Comment by 7c6f434c** [comment]
Built locally including libreoffice-still and libreoffice-fresh?

I agree that this reduces the amount of Qt hacks (manually editing includes is not clean either), so if it doesn't break what works, I am fine with it. (Feel free to self-merge)

I am not exactly happy with the fact that nixpkgs-fmt makes the argument list a few screens long, but apparently someone tries to run nixpkgs-fmt from time to time anyway, so eventually this ladder will sneak in, and it can as well be now.

**Comment by peterhoeg** [comment]
> Built locally including libreoffice-still and libreoffice-fresh?

-qt builds (and runs) locally. -fresh and -still are currently building (which takes *forever* on my hardware).

> I agree that this reduces the amount of Qt hacks (manually editing includes is not clean either), so if it doesn’t break what works, I am fine with it. (Feel free to
> self-merge)

OK.

> I am not exactly happy with the fact that nixpkgs-fmt makes the argument list a few screens long, but apparently someone tries to run nixpkgs-fmt from time
> to time anyway, so eventually this ladder will sneak in, and it can as well be now.

I don’t particularly like the “one argument per line” aspect of nixpkgs-fmt, but I definitely dislike it less than having style discussions based on preference. IMHO, an ugly standard formatter is still better than not having one.

Will wait for the builds to finish here and then merge if otherwise OK. Thank you for reviewing!

**Comment by 7c6f434c** [comment]
>> Built locally including libreoffice-still and libreoffice-fresh?
>
>-qt builds (and runs) locally. -fresh and -still are currently building (which takes *forever* on my hardware).

So it does on mine, but I am also on the go now…

>> I am not exactly happy with the fact that nixpkgs-fmt makes the argument list a few screens long, but apparently someone tries to run nixpkgs-fmt from time
>> to time anyway, so eventually this ladder will sneak in, and it can as well be now.
>
>I don’t particularly like the “one argument per line” aspect of nixpkgs-fmt, but I definitely dislike it less than having style discussions based on preference. IMHO, an ugly standard formatter is still better than not having one.

I am indifferent to ugly and I think that the proper policy for many places is written code style, and whatever is allowed goes if the submitter can say honestly they find this the most readable option. But this approach is apparently «in the wrong direction» to be adopted by Nixpkgs.

I think that splitting the argument list like that goes from ugly to bad usability around 25-line cutoff, but well, navigating LibreOffice expressions is complicated either way, so it doesn't really make things much worse, things are bad on their own.

>Will wait for the builds to finish here and then merge if otherwise OK. Thank you for reviewing!

Thanks for the fixes!

File: pkgs/applications/office/libreoffice/default.nix [hunk 1]
```diff
@@ -278,19 +383,18 @@ in (mkDrv rec {
...
+  '' + optionalString kdeIntegration ''
+    for prog in $out/bin/*; do
+      wrapQtApp $prog
+    done
   '';
 
   dontWrapQtApps = true;
 
   configureFlags = [
     (if withHelp then "" else "--without-help")
```
Comment (SuperSandro2000) [inline pkgs/applications/office/libreoffice/default.nix:395]:
```suggestion
  configureFlags = lib.optional (!withHelp) "--without-help";
    ++ [
```

**Comment by peterhoeg** [comment]
>> +  inherit (lib)
> +    flatten flip
> +    concatMapStrings concatMapStringsSep concatStringsSep
> +    getDev getLib
> +    optional optionals optionalString;
>
> Please revert this. Maintaining this list instead of writing lib. is not an improvement.

We have 3 ways we can deal with stuff in lib:

1. `with lib;` to just pull everything in
2. not pulling in anything but instead using `lib.whatEver`
3. explicitly declaring what we use by `inherit (lib) whatEver;`

Now, option 1 is terrible but option 3 is superior for the following reason: it explicitly lists what we are pulling in for the same reason that we don’t just pass `pkgs` but list what else is needed. It’s closer to the principle of least surprise.

Now, you may prefer to not do it that way. And that’s OK of course. But in the absence of any formal style guide that covers this *and* the fact that the package maintainer doesn’t have an issue with it, we’re just going to leave it as-is.
>
> +  buildInputs = with xorg; [
>
> Can we move with xorg to xorg packages only?

Good point.

> +  configureFlags = lib.optional (!withHelp) “–without-help”;
> +    ++ [

Good catch

Rebuilding libreoffice is a pain, so I think the 2nd and 3rd points should just be addressed when this package is next bumped.

**Comment by 7c6f434c** [comment]
>>> +  inherit (lib)
>> +    flatten flip
>> +    concatMapStrings concatMapStringsSep concatStringsSep
>> +    getDev getLib
>> +    optional optionals optionalString;
>>
>> Please revert this. Maintaining this list instead of writing lib. is not an improvement.
>
>We have 3 ways we can deal with stuff in lib:
>
>1. `with lib;` to just pull everything in
>2. not pulling in anything but instead using `lib.whatEver`
>3. explicitly declaring what we use by `inherit (lib) whatEver;`
>
>Now, option 1 is terrible but option 3 is superior for the following reason: it explicitly lists what we are pulling in for the same reason that we don’t just pass `pkgs` but list what else is needed. It’s closer to the principle of least surprise.
>
>Now, you may prefer to not do it that way. And that’s OK of course. But in the absence of any formal style guide that covers this *and* the fact that the package maintainer doesn’t have an issue with it, we’re just going to leave it as-is.
>>
>> +  buildInputs = with xorg; [
>>
>> Can we move with xorg to xorg packages only?
>
>Good point.
>
>> +  configureFlags = lib.optional (!withHelp) “–without-help”;
>> +    ++ [
>
>Good catch
>
>Rebuilding libreoffice is a pain, so I think the 2nd and 3rd points should just be addressed when this package is next bumped.

Presumably point 2 is not changing the output path? If that change is applied, then if 3 changes the output path, maybe a comment would give a chance not to forget by the next bump

**Comment by SuperSandro2000** [comment]
> Now, option 1 is terrible but option 3 is superior for the following reason: it explicitly lists what we are pulling in for the same reason that we don’t just pass `pkgs` but list what else is needed. It’s closer to the principle of least surprise.

You can't compare pkgs and lib like this. We don't pass in pkgs to easily override/change/switch a package without merging the entire pkgs again and to enable cross-compilation. We don't need to do this or something similar with lib.

**Comment by peterhoeg** [comment]
Of course I can. What you're saying is absolutely correct, but that doesn't in any way invalidate what I'm saying.

And the point is still that you have a preference. Which is fine. I have another. Which is also fine. And barring any official style guide, I'm going with what I like as long as the maintainer of this particular piece of software agrees too.

If you have a *technical* reason, then we can absolutely revisit this but as long as we're in "preference territory" then I'm not really that interested in discussing it as it will lead nowhere.

**Comment by peterhoeg** [comment]
The `xorg` change also causes a rebuild. I suggest we leave as-is.

**Comment by 7c6f434c** [comment]
>> Built locally including libreoffice-still and libreoffice-fresh?
>
>-qt builds (and runs) locally. -fresh and -still are currently building (which takes *forever* on my hardware).

So it does on mine, but I am also on the go now…

>> I am not exactly happy with the fact that nixpkgs-fmt makes the argument list a few screens long, but apparently someone tries to run nixpkgs-fmt from time
>> to time anyway, so eventually this ladder will sneak in, and it can as well be now.
>
>I don’t particularly like the “one argument per line” aspect of nixpkgs-fmt, but I definitely dislike it less than having style discussions based on preference. IMHO, an ugly standard formatter is still better than not having one.

I am indifferent to ugly and I think that the proper policy for many places is written code style, and whatever is allowed goes if the submitter can say honestly they find this the most readable option. But this approach is apparently «in the wrong direction» to be adopted by Nixpkgs.

I think that splitting the argument list like that goes from ugly to bad usability around 25-line cutoff, but well, navigating LibreOffice expressions is complicated either way, so it doesn't really make things much worse, things are bad on their own.

>Will wait for the builds to finish here and then merge if otherwise OK. Thank you for reviewing!

Thanks for the fixes!

**Pull request merged**